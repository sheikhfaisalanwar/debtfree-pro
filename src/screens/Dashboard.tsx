import React, { useState, useEffect } from 'react';
import {
  View,
  Text,
  StyleSheet,
  ScrollView,
  SafeAreaView,
} from 'react-native';
import { DebtCard } from '../components/DebtCard';
import { ProgressBar } from '../components/ProgressBar';
import { StatCard } from '../components/StatCard';
import { Debt } from '../types/Debt';
import { PayoffStrategy } from '../types/Strategy';
import { DebtService } from '../services/DebtService';
import { DataStoreService } from '../services/DataStoreService';

export const Dashboard: React.FC = () => {
  const [debts, setDebts] = useState<Debt[]>([]);
  const [strategy, setStrategy] = useState<PayoffStrategy | null>(null);
  const [extraPayment, setExtraPayment] = useState<number>(100);
  const [loading, setLoading] = useState<boolean>(true);

  useEffect(() => {
    loadData();
  }, []);

  const loadData = async () => {
    try {
      setLoading(true);
      const loadedDebts = await DataStoreService.getDebts();
      const settings = await DataStoreService.getSettings();
      
      setDebts(loadedDebts);
      setExtraPayment(settings.extraPayment);
      
      if (loadedDebts.length > 0) {
        const snowballStrategy = DebtService.calculateSnowballStrategy(loadedDebts, settings.extraPayment);
        setStrategy(snowballStrategy);
      }
    } catch (error) {
      console.error('Failed to load data:', error);
    } finally {
      setLoading(false);
    }
  };

  const formatCurrency = (amount: number): string => {
    return new Intl.NumberFormat('en-CA', {
      style: 'currency',
      currency: 'CAD',
    }).format(amount);
  };

  const totalDebt = debts.reduce((sum, debt) => sum + debt.balance, 0);
  const totalMinPayments = debts.reduce((sum, debt) => sum + debt.minimumPayment, 0);

  if (loading) {
    return (
      <SafeAreaView style={styles.container}>
        <View style={styles.loadingContainer}>
          <Text style={styles.loadingText}>Loading your debt information...</Text>
        </View>
      </SafeAreaView>
    );
  }

  return (
    <SafeAreaView style={styles.container}>
      <ScrollView style={styles.scrollView} showsVerticalScrollIndicator={false}>
        {/* Header */}
        <View style={styles.header}>
          <Text style={styles.title}>ðŸ’³ DebtFree Pro</Text>
          <Text style={styles.subtitle}>Smart debt payoff using the proven snowball method</Text>
        </View>

        {/* Progress Section */}
        <View style={styles.card}>
          <Text style={styles.cardTitle}>ðŸŽ¯ Your Progress</Text>
          <ProgressBar progress={23} label="Overall Progress" />
          
          <View style={styles.statsRow}>
            <StatCard
              value={formatCurrency(totalDebt)}
              label="Total Debt"
              icon="ðŸ’°"
            />
            <StatCard
              value={formatCurrency(totalMinPayments)}
              label="Monthly Payments"
              icon="ðŸ“…"
            />
          </View>
          
          <View style={styles.statsRow}>
            <StatCard
              value="3.2 years"
              label="Debt-Free Date"
              icon="ðŸŽ¯"
            />
            <StatCard
              value="$3,420"
              label="Interest Saved"
              icon="ðŸ’¾"
            />
          </View>
        </View>

        {/* Debts Section */}
        <View style={styles.card}>
          <Text style={styles.cardTitle}>ðŸ’³ Your Debts</Text>
          {debts.map((debt, index) => (
            <DebtCard
              key={debt.id}
              debt={debt}
              isPriority={index === 0} // First debt in snowball method
            />
          ))}
        </View>

        {/* Next Steps */}
        {strategy && (
          <View style={[styles.card, styles.nextStepsCard]}>
            <Text style={styles.nextStepsTitle}>ðŸš€ Next Steps</Text>
            <View style={styles.stepsList}>
              {strategy.debts.length > 0 && (
                <>
                  <Text style={styles.stepText}>
                    1. <Text style={styles.stepBold}>Pay {formatCurrency(strategy.debts[0].monthlyPayment)} toward {debts.find(d => d.id === strategy.debts[0].debtId)?.name}</Text> 
                    {` (minimum ${formatCurrency(debts.find(d => d.id === strategy.debts[0].debtId)?.minimumPayment || 0)} + extra ${formatCurrency(extraPayment)})`}
                  </Text>
                  <Text style={styles.stepText}>
                    2. Pay minimums on all other debts ({formatCurrency(totalMinPayments - (debts.find(d => d.id === strategy.debts[0].debtId)?.minimumPayment || 0))} total)
                  </Text>
                  <Text style={styles.stepText}>
                    3. Expected payoff: {debts.find(d => d.id === strategy.debts[0].debtId)?.name} by {strategy.debts[0].payoffDate.toLocaleDateString()}
                  </Text>
                  {strategy.debts.length > 1 && (
                    <Text style={styles.stepText}>
                      4. After payoff, apply that payment to {debts.find(d => d.id === strategy.debts[1].debtId)?.name}
                    </Text>
                  )}
                </>
              )}
            </View>
          </View>
        )}
      </ScrollView>
    </SafeAreaView>
  );
};

const styles = StyleSheet.create({
  container: {
    flex: 1,
    backgroundColor: '#f3f4f6',
  },
  scrollView: {
    flex: 1,
  },
  header: {
    backgroundColor: 'white',
    padding: 24,
    marginBottom: 16,
    borderBottomWidth: 1,
    borderBottomColor: '#e5e7eb',
  },
  title: {
    fontSize: 28,
    fontWeight: '700',
    color: '#1f2937',
    marginBottom: 4,
  },
  subtitle: {
    fontSize: 16,
    color: '#6b7280',
  },
  card: {
    backgroundColor: 'white',
    margin: 16,
    marginTop: 0,
    padding: 20,
    borderRadius: 16,
    shadowColor: '#000',
    shadowOffset: { width: 0, height: 2 },
    shadowOpacity: 0.1,
    shadowRadius: 8,
    elevation: 5,
  },
  cardTitle: {
    fontSize: 20,
    fontWeight: '600',
    color: '#1f2937',
    marginBottom: 16,
  },
  statsRow: {
    flexDirection: 'row',
    marginBottom: 12,
    marginHorizontal: -4,
  },
  nextStepsCard: {
    backgroundColor: 'rgba(16, 185, 129, 0.05)',
    borderWidth: 1,
    borderColor: 'rgba(16, 185, 129, 0.2)',
  },
  nextStepsTitle: {
    fontSize: 18,
    fontWeight: '600',
    color: '#047857',
    marginBottom: 16,
  },
  stepsList: {
    marginLeft: 8,
  },
  stepText: {
    fontSize: 14,
    color: '#065f46',
    marginBottom: 8,
    lineHeight: 20,
  },
  stepBold: {
    fontWeight: '600',
  },
  loadingContainer: {
    flex: 1,
    justifyContent: 'center',
    alignItems: 'center',
    backgroundColor: '#f3f4f6',
  },
  loadingText: {
    fontSize: 16,
    color: '#6b7280',
    textAlign: 'center',
  },
});