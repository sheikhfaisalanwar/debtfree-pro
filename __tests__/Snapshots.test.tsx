/**
 * @format
 */

import React from 'react';
import ReactTestRenderer from 'react-test-renderer';
import { Dashboard } from '../src/screens/Dashboard';

test('Dashboard snapshot', async () => {
  let tree: ReactTestRenderer.ReactTestRenderer;
  
  await ReactTestRenderer.act(() => {
    tree = ReactTestRenderer.create(<Dashboard />);
  });

  // @ts-ignore
  expect(tree.toJSON()).toMatchSnapshot();
});