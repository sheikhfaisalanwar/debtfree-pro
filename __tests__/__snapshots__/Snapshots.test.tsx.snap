// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Dashboard snapshot 1`] = `
<RCTSafeAreaView
  style={
    {
      "backgroundColor": "#f3f4f6",
      "flex": 1,
    }
  }
>
  <RCTScrollView
    showsVerticalScrollIndicator={false}
    style={
      {
        "flex": 1,
      }
    }
  >
    <View>
      <View
        style={
          {
            "backgroundColor": "white",
            "borderBottomColor": "#e5e7eb",
            "borderBottomWidth": 1,
            "marginBottom": 16,
            "padding": 24,
          }
        }
      >
        <Text
          style={
            {
              "color": "#1f2937",
              "fontSize": 28,
              "fontWeight": "700",
              "marginBottom": 4,
            }
          }
        >
          ðŸ’³ DebtFree Pro
        </Text>
        <Text
          style={
            {
              "color": "#6b7280",
              "fontSize": 16,
            }
          }
        >
          Smart debt payoff using the proven snowball method
        </Text>
      </View>
      <View
        style={
          {
            "backgroundColor": "white",
            "borderRadius": 16,
            "elevation": 5,
            "margin": 16,
            "marginTop": 0,
            "padding": 20,
            "shadowColor": "#000",
            "shadowOffset": {
              "height": 2,
              "width": 0,
            },
            "shadowOpacity": 0.1,
            "shadowRadius": 8,
          }
        }
      >
        <Text
          style={
            {
              "color": "#1f2937",
              "fontSize": 20,
              "fontWeight": "600",
              "marginBottom": 16,
            }
          }
        >
          ðŸŽ¯ Your Progress
        </Text>
        <View
          style={
            {
              "marginBottom": 16,
            }
          }
        >
          <View
            style={
              {
                "flexDirection": "row",
                "justifyContent": "space-between",
                "marginBottom": 8,
              }
            }
          >
            <Text
              style={
                {
                  "color": "#1f2937",
                  "fontSize": 16,
                  "fontWeight": "600",
                }
              }
            >
              Overall Progress
            </Text>
            <Text
              style={
                {
                  "color": "#6b7280",
                  "fontSize": 14,
                }
              }
            >
              23
              % Complete
            </Text>
          </View>
          <View
            style={
              {
                "backgroundColor": "#e5e7eb",
                "borderRadius": 4,
                "height": 8,
                "overflow": "hidden",
              }
            }
          >
            <View
              style={
                [
                  {
                    "backgroundColor": "#10b981",
                    "borderRadius": 4,
                    "height": "100%",
                  },
                  {
                    "width": "23%",
                  },
                ]
              }
            />
          </View>
        </View>
        <View
          style={
            {
              "flexDirection": "row",
              "marginBottom": 12,
              "marginHorizontal": -4,
            }
          }
        >
          <View
            style={
              {
                "alignItems": "center",
                "backgroundColor": "rgba(255, 255, 255, 0.9)",
                "borderColor": "rgba(255, 255, 255, 0.3)",
                "borderRadius": 12,
                "borderWidth": 1,
                "flex": 1,
                "marginHorizontal": 4,
                "padding": 16,
              }
            }
          >
            <Text
              style={
                {
                  "fontSize": 24,
                  "marginBottom": 8,
                }
              }
            >
              ðŸ’°
            </Text>
            <Text
              style={
                {
                  "color": "#1f2937",
                  "fontSize": 20,
                  "fontWeight": "700",
                  "marginBottom": 4,
                }
              }
            >
              $0.00
            </Text>
            <Text
              style={
                {
                  "color": "#6b7280",
                  "fontSize": 12,
                  "textAlign": "center",
                }
              }
            >
              Total Debt
            </Text>
          </View>
          <View
            style={
              {
                "alignItems": "center",
                "backgroundColor": "rgba(255, 255, 255, 0.9)",
                "borderColor": "rgba(255, 255, 255, 0.3)",
                "borderRadius": 12,
                "borderWidth": 1,
                "flex": 1,
                "marginHorizontal": 4,
                "padding": 16,
              }
            }
          >
            <Text
              style={
                {
                  "fontSize": 24,
                  "marginBottom": 8,
                }
              }
            >
              ðŸ“…
            </Text>
            <Text
              style={
                {
                  "color": "#1f2937",
                  "fontSize": 20,
                  "fontWeight": "700",
                  "marginBottom": 4,
                }
              }
            >
              $0.00
            </Text>
            <Text
              style={
                {
                  "color": "#6b7280",
                  "fontSize": 12,
                  "textAlign": "center",
                }
              }
            >
              Monthly Payments
            </Text>
          </View>
        </View>
        <View
          style={
            {
              "flexDirection": "row",
              "marginBottom": 12,
              "marginHorizontal": -4,
            }
          }
        >
          <View
            style={
              {
                "alignItems": "center",
                "backgroundColor": "rgba(255, 255, 255, 0.9)",
                "borderColor": "rgba(255, 255, 255, 0.3)",
                "borderRadius": 12,
                "borderWidth": 1,
                "flex": 1,
                "marginHorizontal": 4,
                "padding": 16,
              }
            }
          >
            <Text
              style={
                {
                  "fontSize": 24,
                  "marginBottom": 8,
                }
              }
            >
              ðŸŽ¯
            </Text>
            <Text
              style={
                {
                  "color": "#1f2937",
                  "fontSize": 20,
                  "fontWeight": "700",
                  "marginBottom": 4,
                }
              }
            >
              3.2 years
            </Text>
            <Text
              style={
                {
                  "color": "#6b7280",
                  "fontSize": 12,
                  "textAlign": "center",
                }
              }
            >
              Debt-Free Date
            </Text>
          </View>
          <View
            style={
              {
                "alignItems": "center",
                "backgroundColor": "rgba(255, 255, 255, 0.9)",
                "borderColor": "rgba(255, 255, 255, 0.3)",
                "borderRadius": 12,
                "borderWidth": 1,
                "flex": 1,
                "marginHorizontal": 4,
                "padding": 16,
              }
            }
          >
            <Text
              style={
                {
                  "fontSize": 24,
                  "marginBottom": 8,
                }
              }
            >
              ðŸ’¾
            </Text>
            <Text
              style={
                {
                  "color": "#1f2937",
                  "fontSize": 20,
                  "fontWeight": "700",
                  "marginBottom": 4,
                }
              }
            >
              $3,420
            </Text>
            <Text
              style={
                {
                  "color": "#6b7280",
                  "fontSize": 12,
                  "textAlign": "center",
                }
              }
            >
              Interest Saved
            </Text>
          </View>
        </View>
      </View>
      <View
        style={
          {
            "backgroundColor": "white",
            "borderRadius": 16,
            "elevation": 5,
            "marginHorizontal": 16,
            "marginVertical": 8,
            "padding": 20,
            "shadowColor": "#000",
            "shadowOffset": {
              "height": 2,
              "width": 0,
            },
            "shadowOpacity": 0.1,
            "shadowRadius": 8,
          }
        }
      >
        <View
          style={
            {
              "marginBottom": 20,
            }
          }
        >
          <Text
            style={
              {
                "color": "#1f2937",
                "fontSize": 20,
                "fontWeight": "600",
                "marginBottom": 8,
              }
            }
          >
            ðŸ“„ Upload Statement
          </Text>
          <Text
            style={
              {
                "color": "#6b7280",
                "fontSize": 14,
                "lineHeight": 20,
              }
            }
          >
            Upload CSV or PDF statements to automatically update your debt information
          </Text>
        </View>
        <View
          accessibilityState={
            {
              "busy": undefined,
              "checked": undefined,
              "disabled": false,
              "expanded": undefined,
              "selected": undefined,
            }
          }
          accessibilityValue={
            {
              "max": undefined,
              "min": undefined,
              "now": undefined,
              "text": undefined,
            }
          }
          accessible={true}
          collapsable={false}
          focusable={true}
          onClick={[Function]}
          onResponderGrant={[Function]}
          onResponderMove={[Function]}
          onResponderRelease={[Function]}
          onResponderTerminate={[Function]}
          onResponderTerminationRequest={[Function]}
          onStartShouldSetResponder={[Function]}
          style={
            {
              "alignItems": "center",
              "backgroundColor": "#3b82f6",
              "borderRadius": 12,
              "justifyContent": "center",
              "marginBottom": 16,
              "minHeight": 80,
              "opacity": 1,
              "padding": 20,
            }
          }
        >
          <View
            style={
              {
                "alignItems": "center",
              }
            }
          >
            <Text
              style={
                {
                  "fontSize": 24,
                  "marginBottom": 8,
                }
              }
            >
              ðŸ“Ž
            </Text>
            <Text
              style={
                {
                  "color": "#ffffff",
                  "fontSize": 16,
                  "fontWeight": "600",
                  "marginLeft": 8,
                }
              }
            >
              Choose Document
            </Text>
            <Text
              style={
                {
                  "color": "#bfdbfe",
                  "fontSize": 12,
                  "marginTop": 4,
                }
              }
            >
              CSV or PDF
            </Text>
          </View>
        </View>
        <View
          style={
            {
              "borderTopColor": "#e5e7eb",
              "borderTopWidth": 1,
              "paddingTop": 16,
            }
          }
        >
          <Text
            style={
              {
                "color": "#374151",
                "fontSize": 14,
                "fontWeight": "600",
                "marginBottom": 8,
              }
            }
          >
            ðŸ“‹ Supported Formats:
          </Text>
          <Text
            style={
              {
                "color": "#6b7280",
                "fontSize": 12,
                "lineHeight": 16,
                "marginBottom": 4,
              }
            }
          >
            â€¢ CSV files with transaction data
          </Text>
          <Text
            style={
              {
                "color": "#6b7280",
                "fontSize": 12,
                "lineHeight": 16,
                "marginBottom": 4,
              }
            }
          >
            â€¢ PDF bank and credit card statements
          </Text>
          <Text
            style={
              {
                "color": "#6b7280",
                "fontSize": 12,
                "lineHeight": 16,
                "marginBottom": 4,
              }
            }
          >
            â€¢ Maximum file size: 10MB
          </Text>
        </View>
      </View>
      <View
        style={
          {
            "backgroundColor": "white",
            "borderRadius": 16,
            "elevation": 5,
            "margin": 16,
            "marginTop": 0,
            "padding": 20,
            "shadowColor": "#000",
            "shadowOffset": {
              "height": 2,
              "width": 0,
            },
            "shadowOpacity": 0.1,
            "shadowRadius": 8,
          }
        }
      >
        <Text
          style={
            {
              "color": "#1f2937",
              "fontSize": 20,
              "fontWeight": "600",
              "marginBottom": 16,
            }
          }
        >
          ðŸ’³ Your Debts
        </Text>
      </View>
    </View>
  </RCTScrollView>
</RCTSafeAreaView>
`;
